components/bpmn/org.wso2.carbon.bpmn.analytics.publisher/src/main/java/org/wso2/carbon/bpmn/analytics/publisher/utils/AnalyticsPublishServiceUtils.java

```diff
@@ -0,0 +1,301 @@
+import org.apache.commons.logging.LogFactory;
+import org.wso2.carbon.bpmn.analytics.publisher.AnalyticsPublisherConstants;
+import org.wso2.carbon.bpmn.analytics.publisher.internal.BPMNAnalyticsHolder;
+import org.wso2.carbon.bpmn.analytics.publisher.models.BPMNProcessInstance;
+import org.wso2.carbon.bpmn.analytics.publisher.models.BPMNTaskInstance;
+
+import org.wso2.carbon.bpmn.core.BPMNServerHolder;
+import org.wso2.carbon.bpmn.core.mgt.model.BPMNVariable;
+import org.wso2.carbon.context.PrivilegedCarbonContext;
+import org.wso2.carbon.context.RegistryType;
+import org.wso2.carbon.registry.api.Registry;
+import org.wso2.carbon.registry.api.RegistryException;
+import org.wso2.carbon.registry.api.Resource;
+
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * AnalyticsPublishServiceUtils is used by the AnalyticsPublisher to fetch the BPMN process instances and the task instances
+ */
+public class AnalyticsPublishServiceUtils {
+	private static Log log = LogFactory.getLog(AnalyticsPublishServiceUtils.class);
+
+	/**
+	 * Get completed process instances which were finished after the given date and time
+	 *
+	 * @return BPMNProcessInstance array if the historic process instance list is not null
+	 */
+	public BPMNProcessInstance[] getCompletedProcessInstances() {
+		HistoryService historyService =
+				BPMNServerHolder.getInstance().getEngine().getHistoryService();
+		HistoricProcessInstanceQuery instanceQuery =
+				historyService.createHistoricProcessInstanceQuery();
+		List<HistoricProcessInstance> historicProcessInstanceList = null;
+		String time = readPublishTimeFromRegistry(AnalyticsPublisherConstants.PROCESS_RESOURCE_PATH,
+		                                          AnalyticsPublisherConstants.LAST_PROCESS_INSTANCE_PUBLISH_TIME);
+		if (time == null) {
+			if (instanceQuery.finished().list().size() != 0) {
+				// if the stored time is null in the registry file then send all completed process instances.
+				historicProcessInstanceList =
+						instanceQuery.finished().orderByProcessInstanceStartTime().asc().list();
+			}
+		} else {
+			Date timeInDateFormat = DateConverter.convertStringToDate(time);
+			int listSize = instanceQuery.finished().startedAfter(timeInDateFormat).list().size();
+			if (listSize != 0) {
+				//avoid to return same object repeatedly if the list has only one object
+				if (listSize == 1) {
+					return null;
+				}
+				//send the process instances which were finished after the given date/time in registry
+				historicProcessInstanceList =
+						instanceQuery.finished().startedAfter(timeInDateFormat)
+						             .orderByProcessInstanceStartTime().asc().listPage(1, listSize);
+			}
+		}
+		if (historicProcessInstanceList != null) {
+			log.debug("Write BPMN process instance to the carbon registry...");
+			writePublishTimeToRegistry(historicProcessInstanceList);
+			//return ProcessInstances set as BPMNProcessInstance array
```

Comment:
i think what we are doing here is to write the last published time (start time of the last process instance) to the registry. We can change the debug log message to mention this. Also add some comments to explain why we need to do this.